mqtt/mqtt-bridge/src/persist/storage/ring_buffer/mod.rs

```diff
@@ -0,0 +1,693 @@
+mod block;
+pub mod error;
+mod mmap;
+
+use crate::persist::{
+    storage::{
+        serialize::{binary_deserialize, binary_serialize},
+        FlushOptions, FlushState, StorageResult,
+    },
+    Key, StorageError,
+};
+use bincode::Result as BincodeResult;
+use block::{
+    calculate_hash, serialized_block_size, validate, BlockHeaderWithHash, Data, BLOCK_HINT,
+};
+use error::RingBufferError;
+use memmap::MmapMut;
+use mmap::{mmap_read, mmap_write};
+use mqtt3::proto::Publication;
+use parking_lot::Mutex;
+use serde::{Deserialize, Serialize};
+use std::io::Result as IOResult;
+use std::{
+    collections::VecDeque,
+    fs::{File, OpenOptions},
+    path::PathBuf,
+    sync::{
+        atomic::{AtomicUsize, Ordering},
+        Arc,
+    },
+    task::Waker,
+    time::Instant,
+};
+
+#[allow(dead_code)]
+fn create_file(file_name: &String) -> IOResult<File> {
+    OpenOptions::new()
+        .read(true)
+        .write(true)
+        .create(true)
+        .open(&PathBuf::from(file_name))
+}
+
+#[allow(dead_code)]
+fn deserialize_block(
+    mmap: Arc<Mutex<MmapMut>>,
+    start: usize,
+    end: usize,
+) -> BincodeResult<BlockHeaderWithHash> {
+    let bytes = mmap_read(mmap.clone(), start, end);
+    binary_deserialize::<BlockHeaderWithHash>(&bytes)
+}
+
+#[allow(dead_code)]
+fn should_flush(flush_options: FlushOptions, flush_state: &FlushState) -> bool {
+    match flush_options {
+        FlushOptions::AfterEachWrite => true,
+        FlushOptions::AfterXWrites(xwrites) => flush_state.writes.load(Ordering::SeqCst) >= xwrites,
+        FlushOptions::AfterXBytes(xbytes) => {
+            flush_state.bytes_written.load(Ordering::SeqCst) >= xbytes
+        }
+        FlushOptions::AfterXMilliseconds(xmillis_elapsed) => {
+            flush_state.millis_elapsed.load(Ordering::SeqCst) >= xmillis_elapsed
+        }
+        FlushOptions::Off => false,
+    }
+}
+
+#[allow(dead_code)]
+fn read_data_maybe_wrap_arond(
+    mmap: Arc<Mutex<MmapMut>>,
+    start: usize,
+    end: usize,
+    file_size: usize,
+) -> Vec<u8> {
+    let mmap = mmap.lock();
+    // Need to wrap around
+    if end > file_size {
+        let file_split = end - file_size;
+        let mut v = vec![];
+        v.append(&mut Vec::from(&mmap[start..file_size]));
```

Comment:
Yes, this part should be optimized to avoid copies. Went with simple for now to be sure I got all the bytes for wrap around case. Can spend some time eliminating copies :)